### PR TITLE
fix: services store register, to restart service informer when switching context

### LINKED ISSUE #6254: subscribe for service/... resource, so informer is started directly when switching context and someone is subscribed
### Bug description

When on the Service page, is the context is switched from this page (using the selector on statusbar for example), the informer is not restarted (before we go to another page and come back to the Service page)

### Operating system

mac os

### Installation Method

Other

### Version

next (development version)

### Steps to reproduce

_No response_

### Relevant log output

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by feloy**
Signed-off-by: Philippe Martin <phmartin@redhat.com>

### What does this PR do?

Add a registry to the ContextsManager to know if some clients have registered to the Services informers, so we can restart the informer automatically when the context is switched by the user. 

### What issues does this PR fix or reference?

Fixes #6254

### How to test this PR?

go to the Services List page, switch the context from the statusbar (without leaving the Service List page), then create a service using kubectl. The Service List should be updated.

- [x] Tests are covering the bug fix or the new feature

### CODE REVIEW

**Review by deboer-tim — CHANGES_REQUESTED** [review]
I confirmed this fixes the original problem, but when testing it is either causing a regression or exposing another one: if I keep selecting the _same_ current context, the resources get shown multiple times. i.e.:

Old behaviour: if I'm on the Services page and select a different context using the status bar the page is empty until I navigate away and come back; if I re-select the context I'm already on nothing changes.

New behaviour: if I change the context the new services appear correctly; if I re-select the current context then 2 services become 4, 6, 8, etc.

File: packages/main/src/plugin/kubernetes-context-state.ts [hunk 1]
```diff
@@ -332,6 +360,14 @@ export class ContextsManager {
       }
     }
 
+    // Restart informers for secondary resources if watchers are subscribing for this resource
+    for (const resourceName of secondaryResources) {
+      if (this.secondaryWatchers.hasSubscribers(resourceName)) {
```
Comment (deboer-tim) [inline packages/main/src/plugin/kubernetes-context-state.ts]: Needs a check like in getCurrentContextResources():
`if (this.secondaryWatchers.hasSubscribers(resourceName) && !this.states.hasInformer(this.currentContext, resourceName)) {`
This fixes the regression I saw while testing.

Comment (feloy) [reply packages/main/src/plugin/kubernetes-context-state.ts]: good catch, pushing a new commit with this fix, thanks

Comment (feloy) [reply packages/main/src/plugin/kubernetes-context-state.ts]: no, moving into this block instead:
```
    if (contextChanged) {

```

**Comment by feloy** [comment]
I still have problems with my tests, they pass in isolation, but fail when ran with the other ones. Checking...

**Review by deboer-tim — APPROVED** [review]
Worked flawlessly in testing, switching back and forth between contexts and making changes.

I really don't like that calling a function like getCurrentContextResources() triggers a background informer, and you have to know to call unregisterGetCurrentContextResources() to unregister it - but the informer was there before and this gives the missing hook to stop it so to me it is most important to get this fix in; I can open an issue or we can discuss later if (e.g.) the functions should be named more like add/remove lIstener.

Needs a rebase since Ingress/Route informers have been merged. kubernetes-contexts-state.ts will need more calls to unregister and mocking like ServiceList.spec.ts might be needed elsewhere or changed.

**Comment by feloy** [comment]
> Worked flawlessly in testing, switching back and forth between contexts and making changes.
> 
> I really don't like that calling a function like getCurrentContextResources() triggers a background informer, and you have to know to call unregisterGetCurrentContextResources() to unregister it - but the informer was there before and this gives the missing hook to stop it so to me it is most important to get this fix in; I can open an issue or we can discuss later if (e.g.) the functions should be named more like add/remove lIstener.
>

Sure, I was also not happy with this name, and the fact that the caller needs to know that it needs to call unregister after this one. After some perspective, I think that registerGetCurrentContextResources would be better. I changed to rename it like this.
 
> Needs a rebase since Ingress/Route informers have been merged. kubernetes-contexts-state.ts will need more calls to unregister and mocking like ServiceList.spec.ts might be needed elsewhere or changed.

Done, thanks

**Review by deboer-tim — APPROVED** [review]
Thanks for function name change and rebasing. Still working flawlessly in testing, so re-approving.

**Review by cdrage — APPROVED** [review]
Was able to:

1. Switch contexts
2. Create a new service
3. It instantly appears in the list.

LGTM code-wise as well as through practical testing.

**Pull request merged**